umbrella-feeds/src/lib.rs

```diff
@@ -0,0 +1,730 @@
+                .0,
+        );
+    }
+
+    Ok(message_hashes)
+}
+
+/// Helper function to verify the signature.
+/// This function throws if the signatures are not valid.
+#[receive(
+    contract = "umbrella_feeds",
+    name = "verifySignatures",
+    parameter = "UpdateParams",
+    crypto_primitives,
+    mutable
+)]
+fn verify_signatures<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    ensure!(
+        param.signers_and_signatures.len() >= host.state().required_signatures as usize,
+        CustomContractError::NotEnoughSignatures
+    );
+
+    let mut prev_signer = AccountAddress([0u8; 32]);
+
+    let message_hashes = contract_view_message_hash(ctx, host, crypto_primitives)?;
+
+    let mut validators: Vec<AccountAddress> = vec![];
+
+    // To save gas we check only required number of signatures.
+    // The case, where you can have part of signatures invalid but still enough valid in total is not supported.
+    for i in 0..host.state().required_signatures {
+        let signer = param.signers_and_signatures[i as usize].0;
+        let signature = &param.signers_and_signatures[i as usize].1;
+
+        //Check signature.
+        let valid_signature =
+            host.check_account_signature(signer, signature, &message_hashes[i as usize])?;
+        ensure!(valid_signature, CustomContractError::WrongSignature);
+
+        ensure!(
+            prev_signer < signer,
+            CustomContractError::SignaturesOutOfOrder
+        );
+
+        validators.push(signer);
+
+        prev_signer = signer;
+    }
+
+    let are_valid_signers = host.invoke_contract_read_only::<Vec<AccountAddress>>(
+        &host.state().staking_bank,
+        &validators,
+        EntrypointName::new_unchecked("verifyValidators"),
+        Amount::zero(),
+    )?;
+
+    let are_valid_signers: bool = are_valid_signers
+        .ok_or(CustomContractError::InvokeContractError)?
+        .get()?;
+
+    ensure_eq!(are_valid_signers, true, CustomContractError::InvalidSigner);
+
+    Ok(())
+}
+
+#[receive(
+    contract = "umbrella_feeds",
+    name = "update",
+    parameter = "UpdateParams",
+    error = "CustomContractError",
+    crypto_primitives,
+    mutable
+)]
+fn update<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> Result<(), CustomContractError> {
+    let param: UpdateParams = ctx.parameter_cursor().get()?;
+
+    let message = param.message;
+
+    // Check that the signature was intended for this contract.
+    ensure_eq!(
+        message.contract_address,
+        ctx.self_address(),
+        CustomContractError::WrongContract
+    );
+
+    // Check signature is not expired.
```

Comment:
It is a feature that we often add to signature verification in smart contracts that would allow you to create a signature and be sure that this signature is only valid until the expiration time. After that, you can be sure that your old signature cannot be used by anyone anymore. 

If you feel that is not useful or rather confusing for your protocol (signature expiration is not so common on Ethereum), I can remove it. 

Another example is where we use signature expiration time.
https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/cis3-nft-sponsored-txs/src/lib.rs#L265

Reply:
does it came with additional cost? If it does, let's remove it, because in our case you can not reuse it anyway once new data will be saved to contract, and you can not use it in other contract or chain because we signing `ChainId` and `target`

Reply:
Expiry is not usually only about reuse. It is also about making sure that when you produce a signature you have certainty it will be used in a fixed amount of time.

The additional cost of this check will be negligible.